statsmodels/tsa/statespace/_initialization.pyx.in

```diff
@@ -0,0 +1,248 @@
+                         int complex_step=False) except 1:
+
+        if initialization_type == 'known':
+            self.initialize_known_constant(offset, initial_state_mean)
+            self.initialize_known_stationary_cov(offset, initial_stationary_state_cov)
+            self.clear_cov(offset, initial_diffuse_state_cov)
+        elif initialization_type == 'diffuse':
+            self.initialize_diffuse(offset, initial_diffuse_state_cov)
+            self.clear_constant(offset, initial_state_mean)
+            self.clear_cov(offset, initial_stationary_state_cov)
+        elif initialization_type == 'approximate_diffuse':
+            self.initialize_known_constant(offset, initial_state_mean)
+            self.initialize_approximate_diffuse(offset, initial_stationary_state_cov)
+            self.clear_cov(offset, initial_diffuse_state_cov)
+        elif initialization_type == 'stationary':
+            self.initialize_stationary_constant(offset, model, initial_state_mean, complex_step)
+            self.initialize_stationary_stationary_cov(offset, model, initial_stationary_state_cov, complex_step)
+            self.clear_cov(offset, initial_diffuse_state_cov)
+        else:
+            raise ValueError('Invalid initialization type')
+
+        return 0
+
+    cdef int clear_constant(self, int offset, {{cython_type}} [:] initial_state_mean) except 1:
+        initial_state_mean[offset:offset + self.k_states] = 0
+        return 0
+
+    cdef int clear_cov(self, int offset, {{cython_type}} [::1, :] cov) except 1:
+        cov[offset:offset + self.k_states, offset:offset + self.k_states] = 0
+        return 0
+
+    cdef int initialize_known_constant(self, int offset,
+                                        {{cython_type}} [:] initial_state_mean) except 1:
+        cdef int inc = 1
+        blas.{{prefix}}copy(&self.k_states, &self.constant[0], &inc,
+                                     &initial_state_mean[offset], &inc)
+
+        return 0
+
+    cdef int initialize_known_stationary_cov(self, int offset,
+                                              {{cython_type}} [::1, :] initial_stationary_state_cov) except 1:
+        cdef int i, inc = 1
+        # Copy columns
+        for i in range(self.k_states):
+            blas.{{prefix}}copy(&self.k_states, &self.stationary_cov[0, i], &inc,
+                                         &initial_stationary_state_cov[offset, offset + i], &inc)
+
+        return 0
+
+    cdef int initialize_diffuse(self, int offset,
+                                 {{cython_type}} [::1, :] initial_diffuse_state_cov) except 1:
+        cdef int i
+        for i in range(offset, offset + self.k_states):
+            initial_diffuse_state_cov[i, i] = 1
+
+        return 0
+
+    cdef int initialize_approximate_diffuse(self, int offset,
+                                             {{cython_type}} [::1, :] initial_stationary_state_cov) except 1:
+        cdef int i
```

Comment:
My understanding is that `Py_ssize_t` is preferred to `int` for this.  You can get away without importing it explicitly or `from cython cimport Py_ssize_t`.

Reply:
That's very interesting, I hadn't seen that before. @bashtage could this be related to the seg faults on appveyor?

Reply:
I'd be surprised.  AFAIK Py_ssize_t is basically an unsigned int.  But its the suggested usage for scalar indexing.